Likes Notification: Display likes using the new endpoints

bitdc opened the pull request:
Refs #15662 

Since the likes data will be fetched from a service, `LikesListController` is introduced to encapsulate the fetching logic. The component mainly serves as a replacement for the table view's delegate and data source in `NotificationDetailsViewController`. Here are a couple of reasons why I chose this approach:

- Minimal impact on the current implementation.
- Minimal effort to implement feature flag (see `setupTableDelegates` method in `NotificationDetailsViewController`).
- Makes it easy to modify later on once we want to implement the Likes page from Reader (e.g. create a thin view controller that uses this component).

By checking out this PR, you should be able to see the look and feel of the new changes directly. However, there are some areas that are not yet covered in this PR:
- [x] Implement feature flag.
- [ ] Add tests for the `LikesListController`. In order to minimize the PR size, I plan to separate the test into another PR.
- [ ] Ensure that the changes do not impact accessibility.
- [ ] Implement Tracks.

Additionally, some review comments from previous PRs were addressed in this change:
- https://github.com/wordpress-mobile/WordPress-iOS/pull/15977#discussion_r585114909
- https://github.com/wordpress-mobile/WordPress-iOS/pull/15994#discussion_r585122810

To test:

- Ensure that the app builds and that you are logged in using an account that already has existing notifications.
- Go to the Notifications tab, and select a notification that's either post likes or comment likes – preferably one that has more than 10 likes. Verify that the list displays all users that liked the content.
- Try to navigate between notifications through the previous/next buttons on the top right navigation bar, and verify that the contents are displayed properly.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

bitdc committed:
Implementation for LikesListController

This component is intended to be used as an alternative table view
delegate and data source, specifically for displaying likes (either for
post, or comment).

bitdc committed:
Add LikesListController in Notification Detail

bitdc committed:
Minor adjustments based on previous review

- https://github.com/wordpress-mobile/WordPress-iOS/pull/15977#discussion_r585114909
- https://github.com/wordpress-mobile/WordPress-iOS/pull/15994#discussion_r585122810

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+
+    let siteID: NSNumber
+
+    let note: Notification?
```
I know this is done in other places in the app, but it would be more clear / less confusing if `note` was the full `notification`.

Also, looks like this can be `private`.

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+        fetchLikes(success: { [weak self] users in
+            if let someUsers = users {
+                self?.likingUsers = someUsers
+            }
```
I think this can just be `self?.likingUsers = users ?? []`.

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+            }
+            self?.isLoadingContent = false
+        }, failure: { [weak self] _ in
+            // TODO: Show error state?
```
👍

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+    private lazy var loadingCell: UITableViewCell = {
+        let cell = UITableViewCell()
+
+        // add activity indicator
```
Nitpick: I don't think this comment is necessary as it's clear what's happening.

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+
+    private let content: ContentIdentifier
+
+    let siteID: NSNumber
```
Looks like this can be `private`.

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+
+    let note: Notification?
+
+    let tableView: UITableView
```
Looks like this can be `private`.

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+
+    private var likingUsers: [RemoteUser] = []
+
+    weak var delegate: LikesListControllerDelegate?
```
Looks like this can be `private`.

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+    func headerCell() -> NoteBlockHeaderTableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: NoteBlockHeaderTableViewCell.reuseIdentifier()) as? NoteBlockHeaderTableViewCell,
+              let group = note?.headerAndBodyContentGroups[Constants.headerRowIndex] else {
+            fatalError()
```
We should just `return` here. Maybe add a `DDLogError` before returning.

If you don't want the return value to be optional, you could just return an `NoteBlockHeaderTableViewCell()`.

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+    func userCell(for indexPath: IndexPath) -> NoteBlockUserTableViewCell {
+        guard indexPath.row < likingUsers.count,
+              let cell = tableView.dequeueReusableCell(withIdentifier: NoteBlockUserTableViewCell.reuseIdentifier()) as? NoteBlockUserTableViewCell else {
+            fatalError()
```
We should just `return` here. Maybe add a `DDLogError` before returning.

If you don't want the return value to be optional, you could just return an `NoteBlockUserTableViewCell
()`.

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListDependency.swift:
```diff
@@ -0,0 +1,22 @@
+
+
+/// A simple, convenient dependency wrapper for LikesListController.
+class LikesListDependency {
```
I'm sorry, I don't quite understand the purpose of this separate class. All its content is only used by `LikesListController`, so why not just define them there?

bitdc replied:
Moved the class into `LikesListController`. Thanks for the suggestion!

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,279 @@
+
+// MARK: - Notification Cell Handling
+
+extension LikesListController {
```
I think this can be a `private` extension since everything in it should be `private`.

ScoutHarris reviewed (changes requested):
Hey David. 

This is looking good. I have a few code comments, suggestions, etc.

Functionally, I did notice one thing. On devices with split view (ex: iPad and iPhone Pro Max), when rotated the table reloads. It should just rotate with the existing content.

bitdc committed:
Add minor changes to address code review

These changes are for https://github.com/wordpress-mobile/WordPress-iOS/pull/16007#pullrequestreview-603373314

bitdc committed:
Add a simple cell for error state

bitdc committed:
Add feature flag for new like notifications

bitdc committed:
Merge branch 'feature/15662-notification_likes' into task/likes-list-controller

bitdc commented:
The latest update should address all of the reviews above, and also:

1. Added a feature flag for the likes notification.
2. Added a simple cell to display when a fetch error occurs.
3. Fixed the reload issue when the device is rotated.

Here's a preview of the error cell:
![Screen Shot 2021-03-04 at 15 17 26](https://user-images.githubusercontent.com/37292612/109932421-bd08b780-7cfc-11eb-8219-025037026132.png)

bitdc committed:
Shorten error cell code

bitdc committed:
minor: tidy up code in LikesListController

bitdc committed:
minor: move static values to constants

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,343 @@
+// MARK: - Dependency Definitions
+
+/// A simple, convenient dependency wrapper for LikesListController.
+class LikesListDependency {
```
Thanks for moving this, but I think my question wasn't clear. I'm confused as to why we have this class period. Its content is simply properties used by `LikesListController`. Why not just make them properties of `LikesListController`?

davidchristiandy replied:
Got it. Initially I thought it would be helpful for unit tests, but I guess I can remove this class instead. Thanks for the input!

bitdc replied:
Ah yes, initially I thought it would be helpful for unit testing later on, but I guess I can remove it to simplify the class. Thanks for the input!

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,343 @@
+    }()
+
+    /**
+     TODO: Update this once we get more updates from the design?
```
Yea, we would do something different here. Can you undo the `errorCell` change please? It'll just have to be removed later. If we do it now while it's fresh, there won't be any confusion later. Thanks!

davidchristiandy replied:
Oh alright, noted on this.

bitdc replied:
Alright, noted! Sorry for this. I think I misunderstood the comment on previous review. 🙏

ScoutHarris replied:
Oh! Sorry to confuse! Thank you!

ScoutHarris commented on WordPress/Classes/ViewRelated/Likes/LikesListController.swift:
```diff
@@ -0,0 +1,343 @@
+        cell.accessoryType = .none
+        cell.name = user.displayName
+
+        // TODO: Re-enable follow functionality once the information is available.
```
🤔 I don't think we are going to re-enable it, correct? Per the design, the follow button on the cell has been removed, and can be accessed via the user's site. So this comment isn't accurate?

bitdc replied:
Oh, got it. I assumed that we wanted to match the previous look first, instead of building it towards the new cell designs directly. Will fix this in the next commit!

ScoutHarris replied:
Yea, generally that was the idea. But I think going ahead as-is is worth not having to add endpoint support for something that would be short lived. Thank you!

bitdc committed:
Remove LikesListDependency

Also addressed some other review comments.

bitdc committed:
Revert errorCell implementation

ScoutHarris reviewed (approved):
I think we got it this time! 😄 Thanks David!
